### Exercise #1

Each year, Inc.--an entrepreneurial-oriented media entity--publishes it's Inc. 5000 List, highlighting the fasting growing startups in the U.S. Attached is a .csv file of the list for 2017. Download the file and perform the following tasks related to data input/output and wrangling.

a. Import the Inc. 5000 data using basic file input techniques (i.e., either via open(filename) or the csv module). Create a dictionary (inc100) that contains the top 100 companies according to the rank (note that the rank is determine based on the growth of the companies revenues from the previous year), using the rank as the key. For each company (rank) in the dictionary, the value should be a dictionary containing the following (populated) fields:

'name': Company name (company)
'founded': Founding date (founded)
'industry': Industry (industry)
'metro': Metropolitan area of company location (metro)
'workers': Number of employees (workers)
Display (print) your full dictionary object and also demonstrate the ability to select specific companies from the dictionary according to their rank (at least two examples).

In [29]:
import csv
import numpy as np
import pandas as pd

In [31]:
path = 'F:/MS in Business Analytics/Spring/Python/Data Files/'
fh = open(path + 'inc5000_2017.csv')
reader = csv.reader(fh)
rank = 0
inc100 = {}
for line in reader :
    if rank == 0 :
        rank += 1
        continue
    elif rank <= 100 :
        inc100[rank] = {'name':line[-1], 'founded':line[-2], 'industry':line[6], 'metro':line[3], 'workers':line[5]}
        rank += 1   
inc100

{1: {'name': 'Skillz',
  'founded': '2012-01-01',
  'industry': 'Software',
  'metro': 'San Francisco',
  'workers': '58'},
 2: {'name': 'EnviroSolar Power',
  'founded': '2012-01-01',
  'industry': 'Energy',
  'metro': 'Dallas',
  'workers': '85'},
 3: {'name': 'Gametime',
  'founded': '2013-01-01',
  'industry': 'Consumer Products & Services',
  'metro': 'San Francisco',
  'workers': '75'},
 4: {'name': 'Club Pilates Franchise',
  'founded': '2007-01-01',
  'industry': 'Consumer Products & Services',
  'metro': 'Los Angeles',
  'workers': '55'},
 5: {'name': 'Halo Top Creamery',
  'founded': '2012-01-01',
  'industry': 'Food & Beverage',
  'metro': 'Los Angeles',
  'workers': '10'},
 6: {'name': 'Small Business Owners of America',
  'founded': '2011-01-01',
  'industry': 'Financial Services',
  'metro': 'Columbus, OH',
  'workers': '15'},
 7: {'name': 'AutomotiveMastermind',
  'founded': '2012-01-01',
  'industry': 'Software',
  'metro': 'New York City',
  'workers': '102'},
 8: {'na

In [32]:
inc100[1]

{'name': 'Skillz',
 'founded': '2012-01-01',
 'industry': 'Software',
 'metro': 'San Francisco',
 'workers': '58'}

In [33]:
inc100[2]

{'name': 'EnviroSolar Power',
 'founded': '2012-01-01',
 'industry': 'Energy',
 'metro': 'Dallas',
 'workers': '85'}

b. Convert your dictionary directly into a DataFrame. What do you observe about the structure of the resultant DataFrame, and how might you modify your process from part a) to adjust this structure into a more appropriate form?

In [34]:
df = pd.DataFrame(inc100)
df

1                  2                             3    \
founded      2012-01-01         2012-01-01                    2013-01-01   
industry       Software             Energy  Consumer Products & Services   
metro     San Francisco             Dallas                 San Francisco   
name             Skillz  EnviroSolar Power                      Gametime   
workers              58                 85                            75   

                                   4                  5    \
founded                     2007-01-01         2012-01-01   
industry  Consumer Products & Services    Food & Beverage   
metro                      Los Angeles        Los Angeles   
name            Club Pilates Franchise  Halo Top Creamery   
workers                             55                 10   

                                       6                     7    \
founded                         2011-01-01            2012-01-01   
industry                Financial Services              Software   
metro                         Columbus, OH         New York City   
name      Small Business Owners of America  AutomotiveMastermind   
workers                                 15                   102   

                              8            9                         10   ...  \
founded                2012-01-01   2012-01-01                2011-01-01  ...   
industry  Advertising & Marketing       Energy        Financial Services  ...   
metro                    San Jose  Los Angeles  Grand Rapids-Wyoming, MI  ...   
name                      Liftoff   Greenspire  FormulaFolio Investments  ...   
workers                        85          150                        43  ...   

                     91                   92            93            94   \
founded       2013-01-01           2008-01-01    2010-01-01    2013-01-01   
industry          Retail  Government Services  Construction   Real Estate   
metro              Miami               Austin       Chicago       Chicago   
name      Shapes Secrets                 MKS2  Hotel Rehabs  Level Office   
workers               65                   58             7            34   

                              95                96                       97   \
founded                2011-01-01        2013-01-01               2013-01-01   
industry  Advertising & Marketing      Construction  Advertising & Marketing   
metro                 Los Angeles         Charlotte           Washington, DC   
name                  Idea Farmer  POWER HOME SOLAR                Harbinger   
workers                        11               109                       21   

                            98             99                            100  
founded              1975-01-01     2010-01-01                    2010-01-01  
industry           Construction         Retail  Consumer Products & Services  
metro        Urban Honolulu, HI  New York City                   Los Angeles  
name      Regal Service Company         S'well                     FabFitFun  
workers                      30             75                            73  

[5 rows x 100 columns]

We observe that there are more columns and less rows as opposed to what we usually observe. It's more like a transpose of what we expect in a data frame.

In order to fix this, we should store all the details as lists inside one single dictionary as opposed to nested dictionaries.

c. Based on your answer from part b), revise your import process from part a) so that you can use the dictionary to create a more appropriately formatted DataFrame. Demonstrate your improved process by displaying (printing) your modified dictionary format and previewing the first 10 rows of your resultant DataFrame.

In [35]:
path = 'F:/MS in Business Analytics/Spring/Python/Data Files/'
fh = open(path + 'inc5000_2017.csv')
reader = csv.reader(fh)
rank = 0
inc100 = {'rank':[], 'name':[], 'founded':[], 'industry':[], 'metro':[], 'workers':[]}
for line in reader :
    if rank == 0 :
        rank += 1
        continue
    if rank <= 100 :
        inc100['rank'].append(rank)
        inc100['name'].append(line[-1])
        inc100['founded'].append(line[-2])
        inc100['industry'].append(line[6])
        inc100['metro'].append(line[3])
        inc100['workers'].append(line[5])
        rank += 1   
print(inc100)

{'rank': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100], 'name': ['Skillz', 'EnviroSolar Power', 'Gametime', 'Club Pilates Franchise', 'Halo Top Creamery', 'Small Business Owners of America', 'AutomotiveMastermind', 'Liftoff', 'Greenspire', 'FormulaFolio Investments', 'Single Serve Beverage Distribution', 'ShipHawk', 'Adva-Net', 'DashBid', 'Forward Financing', 'Solera Specialty Pharmacy', 'Strike Social', 'Reverb.com', 'Excel Impact', 'DGC International', 'Shofur', 'GrowthPlay', 'CalCom Solar', 'ION Solar', 'The Penny Hoarder', 'Matterport', 'Legend Solar', 'SaviLinx', 'TechAnax', 'Spar Information Systems', 'Big Block Realty', 'RelationEdge', 'G

In [36]:
df = pd.DataFrame(inc100)
df.set_index('rank', inplace=True)
df.head(10)

name     founded  \
rank                                                 
1                               Skillz  2012-01-01   
2                    EnviroSolar Power  2012-01-01   
3                             Gametime  2013-01-01   
4               Club Pilates Franchise  2007-01-01   
5                    Halo Top Creamery  2012-01-01   
6     Small Business Owners of America  2011-01-01   
7                 AutomotiveMastermind  2012-01-01   
8                              Liftoff  2012-01-01   
9                           Greenspire  2012-01-01   
10            FormulaFolio Investments  2011-01-01   

                          industry                     metro workers  
rank                                                                  
1                         Software             San Francisco      58  
2                           Energy                    Dallas      85  
3     Consumer Products & Services             San Francisco      75  
4     Consumer Products & Services               Los Angeles      55  
5                  Food & Beverage               Los Angeles      10  
6               Financial Services              Columbus, OH      15  
7                         Software             New York City     102  
8          Advertising & Marketing                  San Jose      85  
9                           Energy               Los Angeles     150  
10              Financial Services  Grand Rapids-Wyoming, MI      43

d. Import the full Inc. 5000 data directly into a DataFrame using an appropriate method, specifying the rank column as the index column. Preview the top 10 fastest growing companies in the data set. After previewing the data, you may notice that the order of the columns is probably not ideal for easily reading the data. For example, the name of the company is in the last column, which should probably be listed first. Reorder the columns of the DataFrame in the following order:

company
founded
leadership
industry
workers
city
metro
state_s
website

Note that you should drop the state_l, url, and id columns. Again, preview the top 10 ranked companies.

In [37]:
df = pd.read_csv(path + 'inc5000_2017.csv', index_col = 'rank')
df.head(10)

state_l state_s           city                     metro  workers  \
rank                                                                         
1     California      CA  San Francisco             San Francisco       58   
2          Texas      TX     Fort Worth                    Dallas       85   
3     California      CA  San Francisco             San Francisco       75   
4     California      CA     Costa Mesa               Los Angeles       55   
5     California      CA    Los Angeles               Los Angeles       10   
6           Ohio      OH       Columbus              Columbus, OH       15   
7       New York      NY  New York City             New York City      102   
8     California      CA      Palo Alto                  San Jose       85   
9     California      CA    Los Angeles               Los Angeles      150   
10      Michigan      MI   Grand Rapids  Grand Rapids-Wyoming, MI       43   

                          industry                               url  \
rank                                                                   
1                         Software                            skillz   
2                           Energy                 envirosolar-power   
3     Consumer Products & Services                          gametime   
4     Consumer Products & Services            club-pilates-franchise   
5                  Food & Beverage                 halo-top-creamery   
6               Financial Services  small-business-owners-of-america   
7                         Software              automotivemastermind   
8          Advertising & Marketing                           liftoff   
9                           Energy                        greenspire   
10              Financial Services          formulafolio-investments   

                              website     id         leadership     founded  \
rank                                                                          
1                   http://skillz.com  70866    Andrew Paradise  2012-01-01   
2         http://envirosolarpower.com  70871       Abraham Issa  2012-01-01   
3                  http://gametime.co  70876      Brad Griffith  2013-01-01   
4              http://clubpilates.com  70881    Anthony Geisler  2007-01-01   
5                  http://halotop.com  70886  Justin Woolverton  2012-01-01   
6                    http://sbooa.com  42948        James Moore  2011-01-01   
7     http://automotivemastermind.com  70891      Marco Schnabl  2012-01-01   
8                   http://liftoff.io  68611         Mark Ellis  2012-01-01   
9           http://greenspirehome.com  42999       David Murray  2012-01-01   
10           http://formulafolios.com  37197         Jason Wenk  2011-01-01   

                               company  
rank                                    
1                               Skillz  
2                    EnviroSolar Power  
3                             Gametime  
4               Club Pilates Franchise  
5                    Halo Top Creamery  
6     Small Business Owners of America  
7                 AutomotiveMastermind  
8                              Liftoff  
9                           Greenspire  
10            FormulaFolio Investments

In [38]:
new_order = ['company','founded','leadership','industry','workers','city','metro','state_s','website']
df = df[new_order]
df.head(10)

company     founded         leadership  \
rank                                                                    
1                               Skillz  2012-01-01    Andrew Paradise   
2                    EnviroSolar Power  2012-01-01       Abraham Issa   
3                             Gametime  2013-01-01      Brad Griffith   
4               Club Pilates Franchise  2007-01-01    Anthony Geisler   
5                    Halo Top Creamery  2012-01-01  Justin Woolverton   
6     Small Business Owners of America  2011-01-01        James Moore   
7                 AutomotiveMastermind  2012-01-01      Marco Schnabl   
8                              Liftoff  2012-01-01         Mark Ellis   
9                           Greenspire  2012-01-01       David Murray   
10            FormulaFolio Investments  2011-01-01         Jason Wenk   

                          industry  workers           city  \
rank                                                         
1                         Software       58  San Francisco   
2                           Energy       85     Fort Worth   
3     Consumer Products & Services       75  San Francisco   
4     Consumer Products & Services       55     Costa Mesa   
5                  Food & Beverage       10    Los Angeles   
6               Financial Services       15       Columbus   
7                         Software      102  New York City   
8          Advertising & Marketing       85      Palo Alto   
9                           Energy      150    Los Angeles   
10              Financial Services       43   Grand Rapids   

                         metro state_s                          website  
rank                                                                     
1                San Francisco      CA                http://skillz.com  
2                       Dallas      TX      http://envirosolarpower.com  
3                San Francisco      CA               http://gametime.co  
4                  Los Angeles      CA           http://clubpilates.com  
5                  Los Angeles      CA               http://halotop.com  
6                 Columbus, OH      OH                 http://sbooa.com  
7                New York City      NY  http://automotivemastermind.com  
8                     San Jose      CA                http://liftoff.io  
9                  Los Angeles      CA        http://greenspirehome.com  
10    Grand Rapids-Wyoming, MI      MI         http://formulafolios.com

e. Upon examining the founding date for each company, you may observe that there are a lot of companies that were 'founded' on January 1. Most likely, these companies were not all founded on New Years Day, but rather there is a lack of information regarding the specific day and month when the company was founded. So rather than list a full date, let's reduce the founding date to just the year, which is the most meaningful (and accurate) unit of time in the data. Using a vectorized approach, convert the 'founded' column into an integer value containing just the year of the founding data. Summarize the frequency of companies that were founded in each year, and describe any trends that you observe in the data. You may want to utilize the following statement so that you can display all of the years in your summary:

pd.set_option('display.max_rows', 500)

In [39]:
df['founded'] = df['founded'].astype(str)
df['founded'] = df['founded'].str[:4].astype(int)
df.head()

company  founded         leadership  \
rank                                                       
1                     Skillz     2012    Andrew Paradise   
2          EnviroSolar Power     2012       Abraham Issa   
3                   Gametime     2013      Brad Griffith   
4     Club Pilates Franchise     2007    Anthony Geisler   
5          Halo Top Creamery     2012  Justin Woolverton   

                          industry  workers           city          metro  \
rank                                                                        
1                         Software       58  San Francisco  San Francisco   
2                           Energy       85     Fort Worth         Dallas   
3     Consumer Products & Services       75  San Francisco  San Francisco   
4     Consumer Products & Services       55     Costa Mesa    Los Angeles   
5                  Food & Beverage       10    Los Angeles    Los Angeles   

     state_s                      website  
rank                                       
1         CA            http://skillz.com  
2         TX  http://envirosolarpower.com  
3         CA           http://gametime.co  
4         CA       http://clubpilates.com  
5         CA           http://halotop.com

In [61]:
pd.set_option('display.max_rows', 500)
df['founded'].value_counts()

2009    445
2012    405
2011    400
2010    390
2008    361
2007    323
2006    245
2005    236
2003    223
2004    217
2001    194
2013    172
2002    165
1999    142
2000    132
1998     97
1997     92
1996     87
1995     60
1994     55
1993     43
1986     41
1992     40
1991     38
1990     34
1989     33
1987     30
1988     25
1983     24
1985     22
1984     19
1982     17
1976     16
1979     14
1981     13
1980     13
1975     10
1978     10
1970      8
1972      8
1977      8
1965      6
1959      5
1956      5
1967      4
1946      4
1964      4
1945      4
1953      3
1960      3
1974      3
1949      3
0         3
1969      3
1948      3
1973      2
1968      2
1951      2
1958      2
1962      2
1947      2
1971      2
1928      1
1908      1
1916      1
1840      1
1952      1
1936      1
1920      1
1895      1
1865      1
1869      1
1950      1
1938      1
1914      1
1902      1
1890      1
1915      1
1919      1
1923      1
1963      1
1961      1
1941      1
1937

We can see that the number of startups have been increasing with time and in particular from the year 2000, it's more than a 100 companies each year.

f. Discretize the number of workers for each company into reasonable categories based on the distribution in the data set. Summarize the frequency of these discretized categories and comment on the size of companies on the Inc. 5000 list.

In [41]:
bins = np.linspace(start=0, stop=5000, num=11)
bins

array([   0.,  500., 1000., 1500., 2000., 2500., 3000., 3500., 4000.,
       4500., 5000.])

In [42]:
pd.cut(df['workers'], bins).value_counts()

(0.0, 500.0]        4666
(500.0, 1000.0]      187
(1000.0, 1500.0]      48
(1500.0, 2000.0]      23
(2000.0, 2500.0]      15
(2500.0, 3000.0]      11
(3500.0, 4000.0]      10
(3000.0, 3500.0]       6
(4500.0, 5000.0]       5
(4000.0, 4500.0]       4
Name: workers, dtype: int64

In [63]:
pd.qcut(df['workers'], 10).value_counts().sort_index()

(-0.001, 12.0]       502
(12.0, 20.0]         597
(20.0, 27.0]         435
(27.0, 36.0]         501
(36.0, 49.0]         472
(49.0, 70.0]         519
(70.0, 100.7]        475
(100.7, 160.0]       502
(160.0, 350.0]       510
(350.0, 140000.0]    489
Name: workers, dtype: int64

From both the above distributions we can clearly see that most of companies in inc5000 have workers less than 500 precisely less than 350.

g.  Summarize the frequency of companies in the data set according to the city where they operate. Print the top 20 cities that are incubators for startup companies. What observations can you make about these areas, and what recommendations would you make about where new startups should locate?

In [44]:
df['city'].value_counts()

New York City        163
Chicago              107
Atlanta              107
Houston               85
Austin                83
San Diego             70
Dallas                56
San Francisco         54
Portland              47
Denver                45
Irvine                44
Tampa                 43
Seattle               43
Washington            42
Los Angeles           42
Boston                41
Phoenix               40
Orlando               36
Indianapolis          30
McLean                30
Arlington             28
Alpharetta            27
Columbus              27
Charlotte             27
Minneapolis           26
Philadelphia          26
Salt Lake City        25
Nashville             25
Cincinnati            25
Scottsdale            25
Alexandria            24
Plano                 24
Huntsville            24
Miami                 23
Brooklyn              23
Louisville            21
Columbia              21
Herndon               21
Las Vegas             20
Reston                20


In [22]:
df['city'].value_counts().head(20)

New York City    163
Chicago          107
Atlanta          107
Houston           85
Austin            83
San Diego         70
Dallas            56
San Francisco     54
Portland          47
Denver            45
Irvine            44
Tampa             43
Seattle           43
Washington        42
Los Angeles       42
Boston            41
Phoenix           40
Orlando           36
Indianapolis      30
McLean            30
Name: city, dtype: int64

All the top 20 incubators for startup companies are the big cities in the United States. So, I recommend that new startups should preferrably locate in these big cities as well.

h.  Explore the types of industries in some of the top 20 cities. Specifically, write a function sum_industry that takes in a DataFrame (df) and a city (default to New York City), and returns the frequency counts of the industry column. Use the function to summarize the industry concentration for New York City, Atlanta, Chicago, and Washington (DC). What observations can you make about the industry concentration in each of these markets?

In [45]:
def sum_industry(df, city='New York City') :
    return df[df['city'] == city]['industry'].value_counts()

In [46]:
sum_industry(df=df, city='New York City')

Advertising & Marketing         41
Business Products & Services    19
IT Services                     17
Human Resources                 12
Health                          10
Consumer Products & Services     9
Software                         8
Financial Services               8
Media                            7
Food & Beverage                  7
Retail                           6
Real Estate                      4
Education                        3
Construction                     3
Travel & Hospitality             3
Insurance                        3
Telecommunications               2
Government Services              1
Name: industry, dtype: int64

In [47]:
sum_industry(df=df, city='Atlanta')

IT Services                     20
Advertising & Marketing         11
Business Products & Services    11
Software                         9
Financial Services               8
Human Resources                  8
Health                           6
Insurance                        4
Consumer Products & Services     4
Logistics & Transportation       4
Real Estate                      3
Food & Beverage                  3
Travel & Hospitality             3
Construction                     3
Retail                           2
Environmental Services           2
Government Services              2
Energy                           1
Security                         1
Telecommunications               1
Engineering                      1
Name: industry, dtype: int64

In [48]:
sum_industry(df=df, city='Chicago')

Advertising & Marketing         17
Business Products & Services    13
IT Services                     11
Real Estate                     11
Software                        10
Financial Services               8
Human Resources                  6
Logistics & Transportation       5
Retail                           5
Consumer Products & Services     5
Health                           4
Energy                           3
Construction                     3
Security                         2
Food & Beverage                  1
Telecommunications               1
Travel & Hospitality             1
Education                        1
Name: industry, dtype: int64

In [49]:
sum_industry(df=df, city='Washington')

Government Services             13
Real Estate                      4
Software                         4
Business Products & Services     4
Advertising & Marketing          4
Consumer Products & Services     3
Energy                           3
Human Resources                  2
IT Services                      2
Financial Services               1
Logistics & Transportation       1
Media                            1
Name: industry, dtype: int64

We can cleary observe that in New York, Atlanta and Chicago, the 3 most dominant industries are Advertising & Marketing, Business Products & Services and IT Services. But in Washington DC although Business Products & Services and Advertising & Marketing are quite dominant, the leading industry is Government Services. 